cli/args/mod.rs

```diff
@@ -742,15 +744,30 @@ pub enum NpmProcessStateKind {
   Byonm,
 }
 
-pub(crate) const NPM_RESOLUTION_STATE_ENV_VAR_NAME: &str =
-  "DENO_DONT_USE_INTERNAL_NODE_COMPAT_STATE";
+pub(crate) const NPM_RESOLUTION_STATE_FD_ENV_VAR_NAME: &str =
+  "DENO_DONT_USE_INTERNAL_NODE_COMPAT_STATE_FD";
 
+// static NPM_PROCESS_STATE: Lazy<Option<NpmProcessState>> = Lazy::new(|| None);
 static NPM_PROCESS_STATE: Lazy<Option<NpmProcessState>> = Lazy::new(|| {
-  let state = std::env::var(NPM_RESOLUTION_STATE_ENV_VAR_NAME).ok()?;
-  let state: NpmProcessState = serde_json::from_str(&state).ok()?;
-  // remove the environment variable so that sub processes
-  // that are spawned do not also use this.
-  std::env::remove_var(NPM_RESOLUTION_STATE_ENV_VAR_NAME);
+  let fd = std::env::var(NPM_RESOLUTION_STATE_FD_ENV_VAR_NAME).ok()?;
+  std::env::remove_var(NPM_RESOLUTION_STATE_FD_ENV_VAR_NAME);
+  let fd = fd.parse::<deno_runtime::deno_io::RawIoHandle>().ok()?;
+  let mut file = {
+    use deno_runtime::deno_io::FromRawIoHandle;
+    unsafe { std::fs::File::from_raw_io_handle(fd) }
+  };
+  let mut buf = Vec::new();
+  // seek to beginning, this file might have been read before
+  file.seek(std::io::SeekFrom::Start(0)).unwrap();
```

Comment:
Would this actually ever happen? It's a unique fd per processs, right? Maybe we shouldn't bother seeking?

Reply:
When the child process inherits the fd, the seek position is shared as well (I spent like an hour debugging why the child was reading an empty string and this was why).

So you have to seek somewhere, you either seek immediately after writing the temp file, or seek before reading.

It does get reused also with lifecycle scripts - there we make a single tempfile and then it gets inherited by all of the lifecycle script subprocesses

Reply:
Ah, I see. Maybe it would be good to say in a comment that the seek is necessary because of the write.

> It does get reused also with lifecycle scripts - there we make a single tempfile and then it gets inherited by all of the lifecycle script subprocesses

Is the position synced between the child processes? If so, there might be a race condition here.

Reply:
It is synced between the child processes, but we don't actually run multiple tasks concurrently (though maybe we should, between packages at least?), so it should be safe. I'll add a comment to that effect